Count and Cache Objectives. (#244)
JuliaManifolds/Manopt.jl: docs/src/tutorials/CountAndCache.md

# How to Cache function calls and Get Started: Optimize!
Ronny Bergmann

In this tutorial, we want to investigate the caching and counting (i.e. statistics) features
of [Manopt.jl](https://manoptjl.org). We will reuse the optimization tasks from the
introductionary tutorial [Get Started: Optimize!](https://manoptjl.org/stable/tutorials/Optimize!.html).

## Introduction

THere are surely many ways to keep track for example of how often the cost function is called,

> for example with a [functor](https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects), as we used in an example in [How to Record Data](https://manoptjl.org/stable/tutorials/HowtoRecord.html)

Author: This should be changed in the `.qmd` file.

``` julia
mutable struct MyCost{I<:Integer}
    count::I
end
MyCost() = MyCost{Int64}(0)
function (c::MyCost)(M, x)
    c.count += 1
    # [ .. Actual implementation of the cost here ]
end
```

though this still leaves a bit of work to the user, especially for trakcing more than just the cost.

> Similarly caching of function calls can for example be added using [Memoize.jl](https://github.com/JuliaCollections/Memoize.jl) by the user.
> The approach in the solvers of [Manopt.jl](https://manoptjl.org) aims to simplify adding
> both these capabilities on the level of calling a solver.

Reviewer: ```suggestion
When the a function like objective or gradient is expensive to compute, it may make sense to cache its results.
Manopt.jl tries to minimize the number of repeated calls but sometimes they are necessary and harmless when the function is cheap to compute.
Caching of expensive function calls can for example be added using [Memoize.jl](https://github.com/JuliaCollections/Memoize.jl) by the user.
The approach in the solvers of [Manopt.jl](https://manoptjl.org) aims to simplify adding
both these capabilities on the level of calling a solver.
```
I think a bit more details as to why a user would want caching is a good idea.

are the [`AbstractManoptProblem`](@ref) and the [`AbstractManoptSolverState`](@ref), where the
former consists of the domain, that is the [manifold](https://juliamanifolds.github.io/ManifoldsBase.jl/stable/types.html#The-AbstractManifold) and [`AbstractManifoldObjective`](@ref).

Both recording and debug capabilities are implemented in a decorator pattern to the solver state.
They can be easily added using the `record=` and `debug=` in any solver call.
This pattern was recently extended, such that also the objective can be decorated.
This is how both caching and counting are implemented, as decorators of the [`AbstractManifoldObjective`](@ref)
and hence for example changing/extending the behaviour of a call to [`get_cost`](@ref).

Let’s finish off the technical background by loading the necessary packages.
Besides [Manopt.jl](https://manoptjl.org) and [Manifolds.jl](https://juliamanifolds.github.io/Manifolds.jl/latest/) we also need
[LRUCaches.jl](https://github.com/JuliaCollections/LRUCache.jl) which are (since Julia 1.9) a weak dependency and provide
the *least recently used* strategy for our caches.

``` julia
using Manopt, Manifolds, Random, LRUCache
```

## Counting

> We first define our task, the Riemannian Center of Mass from the [Get Started: Optimize!](https://manoptjl.org/stable/tutorials/Optimize!.html). tutorial

Reviewer: ```suggestion
We first define our task, the Riemannian Center of Mass from the [Get Started: Optimize!](https://manoptjl.org/stable/tutorials/Optimize!.html) tutorial.
```

```

    # Solver state for `Manopt.jl`s Gradient Descent
    After 67 iterations

    ## Parameters
    * retraction method: ExponentialRetraction()

    ## Stepsize
    ArmijoLineseach() with keyword parameters
      * initial_stepsize    = 1.0
      * retraction_method   = ExponentialRetraction()
      * contraction_factor  = 0.95
      * sufficient_decrease = 0.1

    ## Stopping Criterion
    Stop When _one_ of the following are fulfilled:
        Max Iteration 200:  not reached
        |grad f| < 1.0e-9: reached
    Overall: reached
    This indicates convergence: Yes

    ## Statistics on function calls
      * :Gradient :  202
      * :Cost     :  278
    on a ManifoldGradientObjective{AllocatingEvaluation}

And we see that statistics are shown in the end. To now also cache these calls,
we can use the `cache=` keyword argument.
Since now both the cache and the count “extend” the functionality of the objective,
the order is important: On the high-level interface, the `count` is treated first, which
means that only actual function calls and not cache look-ups are counted.
With the proper initialisation, you can use any caches here that support the

> `get!(function, cache, key)! update. All parts of the objective that can currently be cached are listed at [`ManifoldCachedObjective`](@ref). The solver call has a keyword`cache`that takes a tuple`(c, vs, n)`of three arguments, where`c`is a symbol for the type of cache,`vs`is a vector of symbols, which calls to cache and`n`is the size of the cache. If the last element is not provided, a suitable default (currently`n=10\`) is used.

Reviewer: ```suggestion
`get!(function, cache, key)!` update. All parts of the objective that can currently be cached are listed at [`ManifoldCachedObjective`](@ref). The solver call has a keyword `cache` that takes a tuple`(c, vs, n)` of three arguments, where `c` is a symbol for the type of cache, `vs` is a vector of symbols, which calls to cache and `n` is the size of the cache. If the last element is not provided, a suitable default (currently`n=10`) is used.
```
